src/System.Buffers.Experimental/System/Buffers/BufferExtensions.cs

```diff
@@ -15,14 +15,14 @@ public static class BufferExtensions
         public static void Pipe(this Transformation transformation, ReadOnlyBytes source, IOutput destination)
         {
             int afterMergeSlice = 0;
-            ReadOnlySpan<byte> remainder;
-            Span<byte> stackSpan;
 
-            unsafe
-            {
-                byte* stackBytes = stackalloc byte[stackLength];
-                stackSpan = new Span<byte>(stackBytes, stackLength);
-            }
+            // TODO: "ReadOnlySpan<byte> remainder = stackalloc byte[0]" would fit better here, 
+            //       but it emits substandard IL, see https://github.com/dotnet/roslyn/issues/21952
+            //
+            // Assign 'remainder' to something formally stack-referring.
+            // The default classification is "returnable, not referring to stack", we want the opposite in this case.
+            ReadOnlySpan<byte> remainder = true ? new ReadOnlySpan<byte>() : stackalloc byte[0];
+            Span<byte> stackSpan = stackalloc byte[stackLength];
```

Comment:
Will this be fixed before we ship?

Reply:
Can we just use ReadOnlySpan<byte>.Empty here?

Reply:
@KrzysztofCwalina - yes, will fix this before shipping. The trivial `stackalloc byte[0]` is emitted "correctly", but the IL is embarrassingly redundant, even worse after JIT-ting.  

Optimizing this case in C# compiler backend will benefit existing code as well, so the the value vs. effort is very favorable here.

Reply:
@shiftylogic - if you use ReadOnlySpan.Empty or any other _not_ stack allocated data to initialize the span you are telling the compiler - "I want to be able to return this from the method or mix with other returnables, please make sure I do not refer to stack data by accident".  
That is generally the case.

Here we have a relatively rare opposite - We are slicing through stack allocated data and store in this variable, but have no intent to return it. 
 `stackalloc byte[0]` will convey to the compiler - "please allow referring to stack data, make sure I can never return this from the method"